system-parachains/asset-hubs/asset-hub-kusama/src/xcm_config.rs

```diff
@@ -52,26 +52,32 @@ use xcm_builder::{
 use xcm_executor::{traits::WithOriginFilter, XcmExecutor};
 
 parameter_types! {
-	pub const KsmLocation: MultiLocation = MultiLocation::parent();
+	pub const KsmLocation: Location = Location::parent();
+	pub const KsmLocationV3: xcm::v3::Location = xcm::v3::Location::parent();
 	pub const RelayNetwork: Option<NetworkId> = Some(NetworkId::Kusama);
 	pub RelayChainOrigin: RuntimeOrigin = cumulus_pallet_xcm::Origin::Relay.into();
-	pub UniversalLocation: InteriorMultiLocation =
-		X2(GlobalConsensus(RelayNetwork::get().unwrap()), Parachain(ParachainInfo::parachain_id().into()));
+	pub UniversalLocation: InteriorLocation =
+		[GlobalConsensus(RelayNetwork::get().unwrap()), Parachain(ParachainInfo::parachain_id().into())].into();
 	pub UniversalLocationNetworkId: NetworkId = UniversalLocation::get().global_consensus().unwrap();
-	pub AssetsPalletIndex: u32 = <Assets as PalletInfoAccess>::index() as u32;
-	pub TrustBackedAssetsPalletLocation: MultiLocation = PalletInstance(AssetsPalletIndex::get() as u8).into();
-	pub ForeignAssetsPalletLocation: MultiLocation =
+
+	pub TrustBackedAssetsPalletIndex: u8 = <Assets as PalletInfoAccess>::index() as u8;
+	pub TrustBackedAssetsPalletLocation: Location = PalletInstance(TrustBackedAssetsPalletIndex::get()).into();
+	pub TrustBackedAssetsPalletLocationV3: xcm::v3::Location = xcm::v3::Junction::PalletInstance(TrustBackedAssetsPalletIndex::get()).into();
```

Comment:
Why do we need these `V3` types for things like pallet locations? Is there a plan to remove them once we are happy enough people are on V4? I don't think we want to leave stuff like this forever.

Reply:
Good, point. I have just integrated the changes from XCMv4 (https://github.com/paritytech/polkadot-sdk/pull/1230) here. As far as I know, the intention was to set `xcm::v3::Location` for `pallet_assets` and `pallet_asset_conversion` for `AssetId`, instead of `xcm::latest::Location`, to have better control (for runtime upgrades and migrations, etc.).

I was able to remove `PoolAssetsPalletLocationV3` and `LocalTeleportableToAssetHubV3`, and minimize the usage of `KsmLocationV3`: [Link to the commit](https://github.com/polkadot-fellows/runtimes/pull/187/commits/ea3ebe4d0f98b25d67c49772a2a727e3f4f675e0#diff-9014df9e412de6f140a1c07de05aa154860701d1e2bd39ba8b385d424e346f88).

`KsmLocationV3` fulfills and utilizes the conversion between `pallet_asset_conversion::Config::AssetKind = xcm::v3::Location` and `xcm::latest::Location`, which is internally used by all `xcm-builder` and `xcm-executor` components.

I will attempt to investigate whether we can also eliminate the need for `KsmLocationV3`. I think it might be possible with the correct utilization of `TryFrom<Location>` and `TryInto<Location>` somehow maybe.

I think we should also revisit these two issues before XCMv5:
https://github.com/paritytech/polkadot-sdk/issues/1130
https://github.com/paritytech/polkadot-sdk/issues/1129

cc: @franciscoaguirre

Reply:
We should also migrate all existing foreign assets to V4.

Reply:
opened https://github.com/polkadot-fellows/runtimes/issues/209 for it